### PR TITLE
Support Union Fields When Serializing

### PR DESCRIPTION
**Pull request by william-price01**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
* Added support for Union Fields while serializing structure.
* Added `UnionField` custom field for `Marshmallow` schema generation.
## Issue ticket number and link
[Task](https://griptape-ai.slack.com/lists/T03DNFP0GTW/F06MPE150BG?record_id=Rec07NLKNDG4X)

### CODE REVIEW

File: griptape/structures/structure.py
Comment (collindutter) [inline griptape/structures/structure.py:1]: Missing the `SerializableMixin`. That's what makes adds `to_dict`/`to_json` and makes use of `metadata={"serializable": True}`

File: griptape/structures/structure.py [hunk 1]
```diff
@@ -194,3 +197,27 @@ def run(self, *args) -> Structure:
...
+        elif isinstance(value, (list, tuple)):  # Handle lists or tuples
+            return [self._serialize(item) for item in value]
+        elif isinstance(value, dict):  # Handle dictionaries
+            return {key: self._serialize(val) for key, val in value.items()}
+        elif isinstance(value, Enum):  # Handle Enum types
+            return value.name  # Return the Enum's name as a string
+        elif callable(value):  # Handle functions or methods
+            return "default_function"
+        else:
+            return repr(value)  # Fallback for other non-serializable objects
```
Comment (collindutter) [inline griptape/structures/structure.py:223]: This functionality is provided by `SerializableMixin`, there is no need for a new method.

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/schemas/base_schema.py [hunk 2]
```diff
@@ -47,34 +49,68 @@ def make_obj(self, data: Any, **kwargs) -> Any:
...
             return field_class(allow_none=optional)
 
+        # Unsupported field type
+        raise KeyError(f"Unsupported field type: {field_class}")
+
+    @classmethod
+    def _handle_list(cls, list_arg: type) -> fields.Field:
+        """Handle fields that are lists, including cases with UnionType elements."""
+        # If the argument in the list is a UnionType, handle it as a union within a list
+        if isinstance(list_arg, UnionType) or get_origin(list_arg) is Union:
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:97]: Let's move this logic to a `_is_union` method.

File: griptape/schemas/union_field.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations  # This helps with type hints for forward declarations
```
Comment (collindutter) [inline griptape/schemas/union_field.py:1]: Can remove comment

File: griptape/schemas/union_field.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+import typing as t
```
Comment (collindutter) [inline griptape/schemas/union_field.py:3]: Better to import the types from the module. I.e.
`from typing import Any...`

File: griptape/schemas/union_field.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+class ExceptionGroupError(MarshmallowUnionError):
+    """Collection of possibly multiple exceptions."""
+
+    def __init__(self, msg: str, errors: t.Any) -> None:
+        self.msg = msg
+        self.errors = errors
+        super().__init__(msg, errors)
+
+
+class MarshmallowUnion(marshmallow.fields.Field):
```
Comment (collindutter) [inline griptape/schemas/union_field.py:23]: Rename class to `Union`

File: griptape/tasks/base_task.py [hunk 6]
```diff
@@ -199,3 +200,29 @@ def full_context(self) -> dict[str, Any]:
...
+        elif isinstance(value, (list, tuple)):  # Handle lists or tuples
+            return [self._serialize(item) for item in value]
+        elif isinstance(value, dict):  # Handle dictionaries
+            return {key: self._serialize(val) for key, val in value.items()}
+        elif isinstance(value, Enum):  # Handle Enum types
+            return value.name  # Return the Enum's name as a string
+        elif callable(value):  # Handle functions or methods
+            return "default_function"
+        else:
+            return repr(value)  # Fallback for other non-serializable objects
```
Comment (collindutter) [inline griptape/tasks/base_task.py:228]: This should be removed, as all this logic lives in `SerializableMixin`

File: tests/unit/schemas/test_base_schema.py [hunk 7]
```diff
@@ -68,13 +69,6 @@ def test_get_field_type_info(self):
         assert BaseSchema._get_field_type_info(Literal["foo"]) == (str, (), False)  # pyright: ignore[reportArgumentType]
         assert BaseSchema._get_field_type_info(Literal[5]) == (int, (), False)  # pyright: ignore[reportArgumentType]
 
-    def test_is_list_sequence(self):
-        assert BaseSchema.is_list_sequence(list)
-        assert not BaseSchema.is_list_sequence(tuple)
-        assert not BaseSchema.is_list_sequence(bytes)
-        assert not BaseSchema.is_list_sequence(str)
-        assert not BaseSchema.is_list_sequence(int)
-
```
Comment (collindutter) [inline tests/unit/schemas/test_base_schema.py:77]: Why was this removed?

File: griptape/tools/base_tool.py [hunk 8]
```diff
@@ -39,13 +40,19 @@ class BaseTool(ActivityMixin, ABC):
...
+    input_memory: Optional[list[TaskMemory]] = field(default=None, kw_only=True, metadata={"Serializable": True})
+    output_memory: Optional[dict[str, list[TaskMemory]]] = field(
+        default=None, kw_only=True, metadata={"Serializable": True}
+    )
     install_dependencies_on_init: bool = field(default=True, kw_only=True)
     dependencies_install_directory: Optional[str] = field(default=None, kw_only=True)
-    verbose: bool = field(default=False, kw_only=True)
-    off_prompt: bool = field(default=False, kw_only=True)
+    verbose: bool = field(default=False, kw_only=True, metadata={"Serializable": True})
+    off_prompt: bool = field(default=False, kw_only=True, metadata={"Serializable": True})
```
Comment (collindutter) [inline griptape/tools/base_tool.py:55]: Should be lowercase `serializable`

File: griptape/schemas/base_schema.py [hunk 9]
```diff
@@ -147,18 +199,25 @@ def _resolve_types(cls, attrs_cls: type) -> None:
...
                 "Usage": Message.Usage,
                 "Structure": Structure,
                 "BaseTokenizer": BaseTokenizer,
                 "ToolAction": ToolAction,
                 "Reference": Reference,
                 "Run": Run,
                 "Sequence": Sequence,
+                "BaseRule": BaseRule,
+                "Rule": Rule,
+                "Ruleset": Ruleset,
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:220]: Save this for the structure PR

File: griptape/schemas/base_schema.py [hunk 10]
```diff
@@ -2,12 +2,14 @@
 
 from abc import ABC
 from collections.abc import Sequence
-from typing import Any, Literal, TypeVar, Union, _SpecialForm, get_args, get_origin
+from typing import Any, Literal, TypeVar, _SpecialForm, get_args, get_origin
+from typing import Union as Union_
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:6]: Let's keep this named `Union`. Rename the other `Union` to `UnionField`.

File: griptape/structures/structure.py [hunk 11]
```diff
@@ -22,10 +22,13 @@
...
+    rulesets: list[Ruleset] = field(factory=list, kw_only=True)
+    rules: list[BaseRule] = field(factory=list, kw_only=True)
+    _tasks: list[Union[BaseTask, list[BaseTask]]] = field(
+        factory=list, kw_only=True, alias="tasks"
+    )
     conversation_memory: Optional[BaseConversationMemory] = field(
-        default=Factory(lambda: ConversationMemory()),
-        kw_only=True,
+        default=Factory(lambda: ConversationMemory()), kw_only=True
     )
```
Comment (collindutter) [inline griptape/structures/structure.py:32]: Seem like these changes should be reverted?

**Review by collindutter — COMMENTED** [review]
Let's see what codecov comes back with but overall looks solid! Nice work.

**Review by collindutter — CHANGES_REQUESTED** [review]

**Comment by collindutter** [comment]
Docs seem to be failing because of weird indention on the docstrings. Can you make them consistent with other docstrings?

File: griptape/structures/structure.py [hunk 12]
```diff
@@ -22,7 +22,7 @@
 @define
 class Structure(ABC, RuleMixin):
     id: str = field(default=Factory(lambda: uuid.uuid4().hex), kw_only=True)
-    _tasks: list[BaseTask | list[BaseTask]] = field(factory=list, kw_only=True, alias="tasks")
+    _tasks: list[Union[BaseTask, list[BaseTask]]] = field(factory=list, kw_only=True, alias="tasks")
```
[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/structures/structure.py:25]: I can branch off of this branch, and make these changes there instead.

File: CHANGELOG.md [hunk 13]
```diff
@@ -11,17 +11,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 ### Changed
 
 - **BREAKING**: `BaseEventListener.publish_event` `flush` argument. Use `BaseEventListener.flush_events()` instead.
 - `_DefaultsConfig.logging_config` and `Defaults.drivers_config` are now lazily instantiated.
+- `griptape.schemas.BaseSchema` now uses `griptape.schemas.UnionField` for `Union` fields.
 - `BaseTask.add_parent`/`BaseTask.add_child` now only add the parent/child task to the structure if it is not already present.
 - `BaseEventListener.flush_events()` to flush events from an Event Listener.
 - `BaseEventListener` no longer requires a thread lock for batching events.
-- Updated `TookitTask` system prompt to retry/fix actions when using native tool calling.
+- Updated `ToolkitTask` system prompt to retry/fix actions when using native tool calling.
```
Comment (william-price01) [inline CHANGELOG.md:26]: this was mispelled `TookitTask`

File: griptape/schemas/base_schema.py [hunk 14]
```diff
@@ -52,28 +53,77 @@ def _get_field_for_type(cls, field_type: type) -> fields.Field | fields.Nested:
...
             field_type: A field type.
         """
+        from enum import Enum
+
         from griptape.schemas.polymorphic_schema import PolymorphicSchema
 
         field_class, args, optional = cls._get_field_type_info(field_type)
 
-        # Resolve TypeVars to their bound type
+        # Handle TypeVars
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:62]: What was the reasoning for changing this comment?

File: griptape/schemas/base_schema.py [hunk 15]
```diff
@@ -52,28 +53,77 @@ def _get_field_for_type(cls, field_type: type) -> fields.Field | fields.Nested:
...
-                return fields.List(cls_or_instance=cls._get_field_for_type(args[0]), allow_none=optional)
+                return cls._handle_list(args[0], optional=optional)
             else:
                 raise ValueError(f"Missing type for list field: {field_type}")
-        else:
-            field_class = cls.DATACLASS_TYPE_MAPPING[field_class]
+        field_class = cls.DATACLASS_TYPE_MAPPING.get(field_class)
+        if field_class is None:
+            raise ValueError(f"Unsupported field type: {field_type}")
+        return field_class(allow_none=optional)
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:87]: Please move back into `else`

File: griptape/schemas/base_schema.py [hunk 16]
```diff
@@ -186,3 +236,7 @@ def is_list_sequence(cls, field_type: type | _SpecialForm) -> bool:
                 return issubclass(field_type, Sequence)
         else:
             return False
+
+    @classmethod
+    def is_union_(cls, field_type: type) -> bool:
+        return field_type is Union or get_origin(field_type) is Union
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:242]: Can you please update this and `is_list_sequence` to be protected methods?

Comment (collindutter) [inline griptape/schemas/base_schema.py:242]: Remove trailing underscore.

File: griptape/schemas/base_schema.py [hunk 17]
```diff
@@ -52,28 +53,77 @@ def _get_field_for_type(cls, field_type: type) -> fields.Field | fields.Nested:
...
+            return fields.Constant(None, allow_none=True)
+
+        if cls.is_union_(field_type):
+            return cls._handle_union(field_type, optional=optional)
+
+        elif attrs.has(field_class):
+            schema = PolymorphicSchema if ABC in field_class.__bases__ else cls.from_attrs_cls
+            return fields.Nested(schema(field_class), allow_none=optional)
+
+        elif isinstance(field_class, type) and issubclass(field_class, Enum):
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:76]: Please create an `_is_enum`

File: griptape/schemas/base_schema.py [hunk 18]
```diff
@@ -52,28 +53,77 @@ def _get_field_for_type(cls, field_type: type) -> fields.Field | fields.Nested:
         Args:
             field_type: A field type.
         """
+        from enum import Enum
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:56]: Move to top of class.

File: griptape/schemas/union_field.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+class ExceptionGroupError(MarshmallowUnionError):
+    """Collection of possibly multiple exceptions."""
+
+    def __init__(self, msg: str, errors: Any) -> None:
+        self.msg = msg
+        self.errors = errors
+        super().__init__(msg, errors)
+
+
+class UnionField(marshmallow.fields.Field):
```
Comment (collindutter) [inline griptape/schemas/union_field.py:23]: Class should still be called `Union` for consistency with `Bytes`. When importing elsewhere you can introduce an alias `from griptape.schemas import Union as UnionField`

File: griptape/schemas/base_schema.py [hunk 20]
```diff
@@ -52,28 +53,77 @@ def _get_field_for_type(cls, field_type: type) -> fields.Field | fields.Nested:
...
+        Args:
+            union_type: The Union Type to handle.
+            optional: Whether the Union can be None.
+
+        Returns:
+            A marshmallow Union field.
+        """
+        candidate_fields = [cls._get_field_for_type(arg) for arg in get_args(union_type) if arg is not type(None)]
+
+        if any(arg is type(None) for arg in get_args(union_type)):
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:121]: Split this out into a separate expression. I.e.
```
optional_args = ...
if optional_args:
```

**Review by collindutter — CHANGES_REQUESTED** [review]
Great work getting test coverage up. Just some minor requests left.

File: griptape/schemas/union_field.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    """Field that accepts any one of multiple fields.
```
Comment (collindutter) [inline griptape/schemas/union_field.py:24]: Please include a link to the source repo https://github.com/adamboche/python-marshmallow-union in the description.

File: griptape/schemas/union_field.py [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+
+    Args:
+        fields: The list of candidate fields to try.
+        reverse_serialize_candidates: Whether to try the candidates in reverse order when serializing.
+    """
+
+    def __init__(
+        self,
+        fields: list[marshmallow.fields.Field],
+        *,  # Force reverse_serialize_candidates to be used as a keyword argument
```
Comment (collindutter) [inline griptape/schemas/union_field.py:36]: Remove comment

File: griptape/schemas/union_field.py [hunk 23]
```diff
@@ -0,0 +1,89 @@
+from __future__ import annotations
```
Comment (collindutter) [inline griptape/schemas/union_field.py:1]: Please copy the [license](https://github.com/adamboche/python-marshmallow-union/blob/master/LICENSE) to the top of this file as a module docstring.

File: griptape/schemas/base_schema.py (hunk 14 above)
Comment (william-price01) [reply griptape/schemas/base_schema.py:62]: I changed the logic for handling TypeVars while working through tests, I will change this back. Just ended up accidentally deleting it.

File: griptape/schemas/base_schema.py [hunk 24]
```diff
@@ -60,20 +62,67 @@ def _get_field_for_type(cls, field_type: type) -> fields.Field | fields.Nested:
...
+
+        if cls.is_union(field_type):
+            return cls._handle_union(field_type, optional=optional)
+
+        elif attrs.has(field_class):
+            schema = PolymorphicSchema if ABC in field_class.__bases__ else cls.from_attrs_cls
+            return fields.Nested(schema(field_class), allow_none=optional)
+
+        elif cls.is_enum(field_type):
+            return fields.String(allow_none=optional)
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:76]: Remove whitespace between `if/elif`s

File: griptape/schemas/union_field.py (hunk 23 above)
Comment (collindutter) [reply griptape/schemas/union_field.py:1]: Implemented as NOTICE

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/schemas/base_schema.py (hunk 15 above)
Comment (collindutter) [reply griptape/schemas/base_schema.py:81]: @william-price01 don't forget this one

File: griptape/schemas/base_schema.py [hunk 25]
```diff
@@ -186,3 +230,11 @@ def is_list_sequence(cls, field_type: type | _SpecialForm) -> bool:
...
         else:
             return False
+
+    @classmethod
+    def is_union(cls, field_type: type) -> bool:
+        return field_type is Union or get_origin(field_type) is Union
+
+    @classmethod
+    def is_enum(cls, field_type: type) -> bool:
+        return isinstance(field_type, type) and issubclass(field_type, Enum)
```
Comment (collindutter) [inline griptape/schemas/base_schema.py:240]: Can you please move these methods (and `is_list_sequence`) above the protected methods?

File: griptape/schemas/base_schema.py (hunk 16 above)
Comment (william-price01) [reply griptape/schemas/base_schema.py]: how would you make these protected methods?

File: griptape/schemas/union_field.py (hunk 21 above)
Comment (collindutter) [reply griptape/schemas/union_field.py:24]: @william-price01 please still add this, even with the `NOTICE`

File: griptape/schemas/base_schema.py (hunk 16 above)
Comment (collindutter) [reply griptape/schemas/base_schema.py]: Add a leading underscore. [More info](https://www.geeksforgeeks.org/access-modifiers-in-python-public-private-and-protected/).

File: griptape/schemas/base_schema.py (hunk 25 above)
Comment (collindutter) [reply griptape/schemas/base_schema.py:240]: Ignore me here, my other comment said to make them protected. Continue with that.

File: griptape/schemas/base_schema.py (hunk 16 above)
Comment (william-price01) [reply griptape/schemas/base_schema.py]: @collindutter nvm figured that out.

File: griptape/schemas/union_field.py (hunk 21 above)
Comment (william-price01) [reply griptape/schemas/union_field.py:24]: Weird, I added this, can you not see the changes?

File: griptape/schemas/union_field.py [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+    def __init__(self, msg: str, errors: Any) -> None:
+        self.msg = msg
+        self.errors = errors
+        super().__init__(msg, errors)
+
+
+class Union(marshmallow.fields.Field):
+    """Field that accepts any one of multiple fields.
+
+    Source: https://github.com/adamboche/python-marshmallow-union
```
Comment (william-price01) [inline griptape/schemas/union_field.py:26]: Added here @collindutter

**Review by collindutter — APPROVED** [review]
Fantastic work!

**Review by collindutter — APPROVED** [review]
